TraitsExpanded/Interfaces/TraitSet.cs

```diff
@@ -0,0 +1,109 @@
+﻿using System.Collections.Generic;
+using TaleWorlds.CampaignSystem;
+using TaleWorlds.Core;
+using TaleWorlds.MountAndBlade;
+
+namespace TraitsExpanded
+{
+    public abstract class TraitSet
+    {
+        public virtual List<ITrait> Traits { get; set; } = new List<ITrait>();
+        public virtual string ID { get; set; }
+        public virtual int CurrentTraitIndex { get; set; } = 0;
+        public virtual NPCRestrictionEnum NPCRestriction { get; set; } = NPCRestrictionEnum.NONE;
+        public virtual bool IsActive { get; set; } = false;
+        public virtual CharacterObject CurrentCharacter { get; set; }
+
+        public virtual void Init(CharacterObject character)
+        {
+            CurrentCharacter = character;
+            Util.LogMessage($"{ID} initialized");
+        }
+
+        public virtual bool Activate()
+        {
+            var succeeded = false;
+
+            if (Traits.Count > 0 && CurrentCharacter != null)
+            {
+                succeeded = Traits[CurrentTraitIndex].Activate(CurrentCharacter);
+                IsActive = true;
+            }
+            else
+            {
+                IsActive = false;
+                succeeded = true;
+            }
+
+            return succeeded;
+        }
+
+        public virtual bool Deactivate()
+        {
+            var succeeded = false;
+
+            if (Traits.Count > 0 && CurrentCharacter != null)
+            {
+                succeeded = Traits[CurrentTraitIndex].Deactivate(CurrentCharacter);
+                IsActive = false;
+            }
+
+            return succeeded;
+        }
+
+        public virtual bool Upgrade()
```

Comment:
the upgrade/downgrade logic seems based around traits that start at 0 and only increase, but don't we want to support trait sets containing both negative and positive traits? e.g. a trait representing trustworthiness:
[-2, -1, 0, 1, 2], where index -2 is named "scheming" and has a cost -10, index -1 is named "suspicious" and has a cost of -5, index 0 is neutral, index 1 is named "trustworthy" and has a cost of 5, index 2 is named "reliable" and has a cost of 10

Reply:
you could represent this with two lists, one for positive and one for negative traits in the set. if the index is negative, its absolute value becomes the index into the negative trait list. if positive, its value is the index into the positive trait list.

Reply:
I'm not sure if going into negative indexes is needed. We can simply start from zero being the worst trait, and 5 (in your example) being the best

Reply:
sure, that works fine. it probably does make sense not to have a built-in idea of whether a trait is positive or negative, as what a trait does or means is arbitrary.